api/src/main/java/io/grpc/CallOptions.java

```diff
@@ -328,7 +354,8 @@ public <T> CallOptions withOption(Key<T> key, T value) {
       }
     }
 
-    newOptions.customOptions = new Object[customOptions.length + (existingIdx == -1 ? 1 : 0)][2];
+    builder.customOptions = new Object[customOptions.length + (existingIdx == -1 ? 1 : 0)][2];
+    CallOptions newOptions = builder.build();
```

Comment:
I guess this and the following code (next 9 lines) is most probably correct but quite confusing. Why not add the new custom option to `builder.customOptions` first and then just return `builder.build()`. So there is no real need to declare `newOptions` and manipulate it